File: backend/src/main/java/com/sipelka/backend/model/Proposal.java
package com.sipelka.backend.model;

import com.github.f4b6a3.uuid.UuidCreator;
import com.sipelka.backend.model.enums.StatusProposal;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "proposal", indexes = {
        @Index(name = "idx_proposal_status", columnList = "status_proposal"),
        @Index(name = "idx_proposal_peneliti", columnList = "peneliti_id"),
        @Index(name = "idx_proposal_created_at", columnList = "created_at")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Proposal {

    @Id
    @Column(name = "proposal_id", updatable = false, nullable = false)
    private UUID id;

    @PrePersist
    protected void onCreate() {
        if (this.id == null) {
            this.id = UuidCreator.getTimeOrderedEpoch();
        }
        if (this.createdAt == null) {
            this.createdAt = LocalDateTime.now();
        }
        if (this.updatedAt == null) {
            this.updatedAt = LocalDateTime.now();
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "peneliti_id", nullable = false)
    private User peneliti;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hibah_id", nullable = false)
    private ProgramHibah hibah;

    @Column(name = "judul_penelitian", nullable = false)
    private String judulPenelitian;

    @Column(name = "bidang_penelitian", nullable = false)
    private String bidangPenelitian;

    @Column(columnDefinition = "TEXT")
    private String ringkasan;

    @Column(name = "dokumen_url", nullable = false)
    private String dokumenUrl;

    @Enumerated(EnumType.STRING)
    @Column(name = "status_proposal")
    private StatusProposal statusProposal;

    @Column(name = "kriteria_kelengkapan_dokumen")
    private Boolean kriteriaKelengkapanDokumen;

    @Column(name = "kesesuaian_bidang")
    private Boolean kesesuaianBidang;

    @Column(name = "skor_rule_based")
    private Integer skorRuleBased;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: front_end_mobile_sipelka/lib/models/proposal.dart
class Proposal {
  final String id;
  final String penelitiId;
  final String? penelitiName;
  final String hibahId;
  final String? hibahName;
  final String judulPenelitian;
  final String bidangPenelitian;
  final String? ringkasan;
  final String? dokumenUrl;
  final String statusProposal;
  final bool? kriteriaKelengkapanDokumen;
  final bool? kesesuaianBidang;
  final int? skorRuleBased;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Proposal({
    required this.id,
    required this.penelitiId,
    this.penelitiName,
    required this.hibahId,
    this.hibahName,
    required this.judulPenelitian,
    required this.bidangPenelitian,
    this.ringkasan,
    this.dokumenUrl,
    required this.statusProposal,
    this.kriteriaKelengkapanDokumen,
    this.kesesuaianBidang,
    this.skorRuleBased,
    this.createdAt,
    this.updatedAt,
  });

  factory Proposal.fromJson(Map<String, dynamic> json) {
    return Proposal(
      id: json['id'] as String,
      penelitiId: json['penelitiId'] as String,
      penelitiName: json['penelitiName'] as String?,
      hibahId: json['hibahId'] as String,
      hibahName: json['hibahName'] as String?,
      judulPenelitian: json['judulPenelitian'] as String,
      bidangPenelitian: json['bidangPenelitian'] as String,
      ringkasan: json['ringkasan'] as String?,
      dokumenUrl: json['dokumenUrl'] as String?,
      statusProposal: json['statusProposal'] as String,
      kriteriaKelengkapanDokumen: json['kriteriaKelengkapanDokumen'] as bool?,
      kesesuaianBidang: json['kesesuaianBidang'] as bool?,
      skorRuleBased: json['skorRuleBased'] as int?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'penelitiId': penelitiId,
    'penelitiName': penelitiName,
    'hibahId': hibahId,
    'hibahName': hibahName,
    'judulPenelitian': judulPenelitian,
    'bidangPenelitian': bidangPenelitian,
    'ringkasan': ringkasan,
    'dokumenUrl': dokumenUrl,
    'statusProposal': statusProposal,
    'kriteriaKelengkapanDokumen': kriteriaKelengkapanDokumen,
    'kesesuaianBidang': kesesuaianBidang,
    'skorRuleBased': skorRuleBased,
    'createdAt': createdAt?.toIso8601String(),
    'updatedAt': updatedAt?.toIso8601String(),
  };
}
